tools/flaky-test-reporter/jsonreport/jsonreport.go

```diff
@@ -58,61 +64,94 @@ func (r *Report) writeToArtifactsDir() error {
+// getLatestValidBuild inexpensively sorts and finds the most recent JSON report.
+// Assumes sequential build IDs are sequential in time.
+func getLatestValidBuild(job *prow.Job, repo string) (int, error) {
+	// check latest build first, before looking to older builds
+	if buildID, err := job.GetLatestBuildNumber(); err == nil {
+		build := job.NewBuild(buildID)
+		if reports := getReportPaths(build, repo); len(reports) != 0 {
+			return buildID, nil
+		}
+	}
+	// look at older builds
 	maxElapsedTime, _ := time.ParseDuration(fmt.Sprintf("%dh", maxAge*24))
 	buildIDs := job.GetBuildIDs()
 	sort.Sort(sort.Reverse(sort.IntSlice(buildIDs)))
 	for _, buildID := range buildIDs {
 		build := job.NewBuild(buildID)
-		// check for JSON
-		if contents, err := build.ReadFile(relPath); err != nil {
+		// check if reports exist for this build
+		if reports := getReportPaths(build, repo); len(reports) == 0 {
+			continue
+		}
+		// check if this report is too old
+		startTimeInt, err := build.GetStartTime()
+		if err != nil {
 			continue
+		}
+		startTime := time.Unix(startTimeInt, 0)
+		if time.Since(startTime) < maxElapsedTime {
+			return buildID, nil
 		} else {
-			// JSON exists, check timestamp
-			startTimeInt, err := build.GetStartTime()
-			if err != nil {
-				continue
-			}
-			startTime := time.Unix(startTimeInt, 0)
-			if time.Since(startTime) < maxElapsedTime {
-				return contents, nil
-			} else {
-				return nil, fmt.Errorf("latest JSON log is outdated: %.2f days old", time.Since(startTime).Hours()/24)
-			}
+			return 0, fmt.Errorf("latest JSON log is outdated: %.2f days old", time.Since(startTime).Hours()/24)
 		}
 	}
-	return nil, fmt.Errorf("no JSON logs found in recent builds")
+	return 0, fmt.Errorf("no JSON logs found in recent builds")
 }
 
-// GetReportForRepo gets a Report struct from a specific build for a given repo
-// use buildID = -1 for most recent successful report
-func GetReportForRepo(repo string, buildID int) (*Report, error) {
-	report := &Report{
-		Repo: repo,
+// getReportPaths searches build artifacts for reports from the given repo, returning
+// the path to any matching files. Use repo = "" to get all reports from all repos.
+func getReportPaths(build *prow.Build, repo string) []string {
+	var matches []string
+	suffix := path.Join(repo, filename)
+	for _, artifact := range build.GetArtifacts() {
+		if strings.HasSuffix(artifact, suffix) {
+			matches = append(matches, artifact)
+		}
 	}
-	contents, err := getJSONForBuild(repo, buildID)
+	return matches
+}
+
+// readJSONReport builds a repo-specific report object from a given json file path.
+func readJSONReport(build *prow.Build, filename string) (*RepoReport, error) {
+	report := &RepoReport{}
+	contents, err := build.ReadFile(filename)
 	if err != nil {
 		return nil, err
 	}
-	if err = json.Unmarshal(contents, report); err != nil {
+	if err := json.Unmarshal(contents, report); err != nil {
```

Comment:
doesn't it complain? err is already defined above

Reply:
surprisingly, no. Seems like it defines a new err within the scope of the if statement. I'll change it for consistency anyways.